doc/sphinx/proof-engine/ltac.rst

```diff
@@ -713,100 +747,183 @@ tactic
-   In this latter case, the definitions are evaluated lazily so that the rec
-   keyword can be used also in non-recursive cases so as to avoid the eager
+   Creates a local definition.
+   Using :tacn:`let` `rec` makes the definition recursive and causes
+   the definitions to be evaluated lazily.  Thus, :tacn:`let` `rec`
+   can be used in non-recursive cases to avoid eager
    evaluation of local definitions.
 
    .. but rec changes the binding!!
 
+   :tacn:`let` is a :token:`binder_tactic`.
+
 Application
 ~~~~~~~~~~~
 
 An application is an expression of the following form:
 
-.. tacn:: @qualid {+ @tacarg}
+.. tacn:: @qualid {* @tactic_arg }
 
    The reference :n:`@qualid` must be bound to some defined tactic definition
-   expecting at least as many arguments as the provided :n:`tacarg`. The
+   expecting at least as many arguments as the provided :n:`tactic_arg`. The
    expressions :n:`@ltac_expr__i` are evaluated to :n:`v__i`, for i = 1, ..., n.
 
    .. what expressions ??
 
-Function construction
-~~~~~~~~~~~~~~~~~~~~~
+Function construction: fun
+~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 A parameterized tactic can be built anonymously (without resorting to
 local definitions) with:
 
-.. tacn:: fun {+ @ident} => @ltac_expr
+.. tacn:: fun {+ @name } => @ltac_expr
 
    Indeed, local definitions of functions are a syntactic sugar for binding
    a :n:`fun` tactic to an identifier.
 
-Pattern matching on terms
-~~~~~~~~~~~~~~~~~~~~~~~~~
+   :tacn:`fun` is a :token:`binder_tactic`.
+
+Pattern matching on terms: match
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. tacn:: @match_key @ltac_expr with {? %| } {+| @match_rule } end
+   :name: match; lazymatch; multimatch
+
+   .. insertprodn match_key cpattern
+
+   .. prodn::
+      match_key ::= match
+      | multimatch
+      | lazymatch
+      match_rule ::= {| @match_pattern | _ } => @ltac_expr
+      match_pattern ::= @cpattern
+      | context {? @ident } [ @cpattern ]
+      cpattern ::= @term
+
+   .. todo I don't fully understand these descriptions.  I think some concepts
+      such as backtracking need to be explained first and that the lanugage
+      can probably be simplified.
+
+   :tacn:`match`, :tacn:`lazymatch` and :tacn:`multimatch` are :token:`ltac_expr1`\s.
+
+   `multimatch`
+      Allows subsequent tactics to
+      backtrack into a right-hand side tactic which has backtracking points
+      left and trigger the selection of a new matching branch when all the
+      backtracking points of the right-hand side have been consumed.
+
+      :tacn:`match` :n:`…` is, in fact, shorthand for :tacn:`once` :tacn:`multimatch` `…`.
+
+   `lazymatch`
+      Causes the match to commit to the first matching branch
+      rather than trying a new matching if the right-hand side fails. If
+      the right-hand side of the selected branch is a tactic with
+      backtracking points, then subsequent failures cause this tactic to
+      backtrack.
```

Comment:
It needs an introductory paragraph covering matches in general, then an entry for `match`.  I think the position will be OK if that's added.  Hard to write something because I don't understand how `match` differs from the other 2.  See the `todo PR` I added.

We need a paragraph or two explaining backtracking, which I don't understand well enough to explain (I can only guess).  It seems to apply to several constructs, such as match, +, tryif, first, so maybe it should be a subsection.  Which ones does it apply to?  "Backtracking branching" is just a different phrase for the same thing?

I guess the difference is relevant for constructs like match ... end + match ... end.

Reply:
Yes, backtracking applies to all this and is probably best explain in the description of `+`. "Backtracking branching" is perhaps a bit confusing for a section title. It could be "Creating backtracking points".

Reply:
I'm matching the behavior of `tac + tac` and `tac ; tac` to their current descriptions.  `+` talks about backtracking, but only `;` does it below.  Does the current doc have it backwards?  What's the story?

```

multimatch False with
| False => idtac "1"
| _ => idtac "2"
end +
multimatch False with
| False => idtac "1B"; fail
| _ => idtac "2B"; fail
end.

(* 1 *)

multimatch False with
| False => idtac "1"
| _ => idtac "2"
end ;
multimatch False with
| False => idtac "1B"; fail
| _ => idtac "2B"; fail
end.

(*
Tactic failure.
1
1B
2B
2
1B
2B
*)
```

Reply:
The difference between `+` and `;` is not backtracking (that's the difference between `match`, `lazymatch`, and `multimatch`).  The difference is that `+` is alternatives, while `;` is sequencing.  If you imagine each tactic produces a potentially infinite stream of successes, together with a particular error to raise when you run out of successes, then the semantics of `a + b` are "concatenate the stream of `a` with the stream of `b`, and if you run out of all of these, use the error message from `b`".  On the other hand, `a; b` is sort-of a flat-map: "for each success of `a` with side effects `A`, for each success of `b` with side effects `B`, have the success with side effect '`A` followed by `B`'".

In your first case, what happens is that your first `multimatch` succeeds, so the second one is never run, because you just end up using the success from the first alternative.  In the second case, with `;`, what happens is that the first tactic has two successes, `idtac "1"` and `idtac "2"`.  So first you `idtac "1"`, and see if you can get a success from the second tactic.  When the second tactic fails, you go back and try the other success of the first one, `idtac "2"`, and then try the second tactic again.

Reply:
I'll need considerable elaboration to make sense of your comment--too many unclear concepts.  Let's take it bit by bit.  Concrete examples are helpful.

- I don't see why `;' isn't considered backtracking.  Certainly in the ordinary sense of the word it is backtracking.  Why is the distinction significant?

- For '+', the doc says "The tactic value v1 is applied to each focused goal independently and if it fails or a later tactic fails, then the proof **backtracks** to the current goal and v2 is applied.".  Can you give me an example for this?

- "Success" seems to be used very loosely in the existing doc.  This is particularly opaque: "If you imagine each tactic produces a potentially infinite stream of successes, together with a particular error to raise when you run out of successes".  What's a concrete example or two of producing a very large number of successes from a single tactic?  (FWIW, an infinite series of successes would take forever to run :-)

Reply:
Please read http://coqhott.gforge.inria.fr/blog/coq-tactic-engine/, an excellent explanation of the tactic engine by @ppedrot.  I'd rather not re-explain various bits of it piecemeal.

I expect your first two points to be addressed by this blog post.  To give you concrete examples for the final bullet point:
> What's a concrete example or two of producing a very large number of successes from a single tactic? (FWIW, an infinite series of successes would take forever to run :-)

If you have an inductive type `T` with `n` constructors, the `constructor` tactic produces `n` successes on the goal `T`.  The tactic `Ltac strange := idtac + strange.` produces infinitely many successes (importantly, the potentially infinite stream is lazy, so you don't run them all when you don't need to); `strange` succeeds instantly, but `strange; fail` runs forever.

Reply:
I read the post.  For the first two, it seems like I'd have to explain some aspects of monads in order to describe the behavior of Ltac, though perhaps we don't need to introduce the term itself--not sure it's essential to know the term to use Ltac, let alone Coq.  Not sure the relevant points can be explained simply and fully.

In a number of places, the doc mentions evaluating ltac expressions as a step before applying them.  Is there a separation of evaluation and application into separate phases, e.g. evaluation corresponds to constructing the monads, the application is execution of the monads?

There is no mention of the Ltac `;` or `+` constructs in the post.

The infinite successes would be a bit fun to explain.  I take it that that's useful though the example is not realistic.

Reply:
> In a number of places, the doc mentions evaluating ltac expressions as a step before applying them. Is there a separation of evaluation and application into separate phases,

Yes, evaluation is what I called "expression evaluation" in https://github.com/coq/coq/pull/12103#discussion_r432990765, and "application" is what I called "runtime".

> e.g. evaluation corresponds to constructing the monads, the application is execution of the monads?

No, again, see https://github.com/coq/coq/pull/12103#discussion_r432990765 for the distinction.


> There is no mention of the Ltac `;` or `+` constructs in the post.

`;` is sequencing in the monad, and corresponds to `tclBIND` in the post when the second tactic takes no argument. The `+` construct almost the same thing as `tclOR` (again when the second tactic takes no argument), except that it runs on each goal separately.  that is, `a + b` is always the same as `[ > a + b .. ]`, while this is not true of `tclOR` (note though that `tclOR` is not exposed to the user of Ltac, but is only exposed to OCaml plugin developers).

Reply:
> The infinite successes would be a bit fun to explain. I take it that that's useful though the example is not realistic.

It's useful to have multiple successes, and it's useful not to limit the number to any particular finite number.  Given arbitrary recursion, it's not practical to enforce that the number of successes is finite.  I've never seen an actual application of infinitely many successes.  The first one that comes to mind is implementing a parser in Ltac on a grammar that admits a potentially infinite number of valid parses (e.g., some ambiguous regular expression grammars).

Reply:
Oh, here's another tactic which is more useful and ends up having an infinite number of successes:
```coq
Ltac inhabit := constructor; inhabit.
```
This tactic will try to inhabit inductive types.  It has an infinite number of successes on `nat`, and you can use this do proof search with something like
```coq
unshelve eexists; [ inhabit | <some sort of proof procedure that will only succeed on some nats> ]
```
(The tactic will loop if you set up your inductive type so that the base case comes after the recursive case, though.)

Reply:
I'll ponder your comments Monday.

It would nice if you could give an example showing backtracking for `+`.  The doc says "The tactic value v1 is applied to each focused goal independently and if it fails or a later tactic fails, then the proof backtracks to the current goal and v2 is applied."

Reply:
Contrived example of backtracking using `+`:
```coq
Goal unit * True.
  split; (pose tt + pose I); assumption.
Qed.
```